### PR TITLE
Implement Advanced Scraping Algorithm

### LINKED ISSUE #43: Add more internship scrape config attributes
- [ ] Restructure scrape config to include actions
- [ ] Add additional attributes to properties (xpath, regex)
- [ ] Add next page button

### PR DESCRIPTION
**Pull request by JeremyEastham**
### Improvements and Changes
- Closes #43
- Renamed `name` in `companies.csv` to `company`
- Created a `WebScraper` class to manage a Selenium instance and provide convenience methods for scraping
- Updated `scraping_config.json`
  + Each company's `scrape` property now contains an array of `action` objects
  + Based on each action's name (the `action` property), the other properties of each object are passed as parameters into functions annotated with `@scrape_action("name")`
  + These parameters are type-checked with the Pydantic library based on model classes (see `scraping.actions.models`)
  + For instance, if a parameter is annotated with Pydantic's `AnyUrl`, it is matched against a regex that validates URLs before being passed into the function
  + I also built a dependency-injection system similar to Angular's that injects instances of classes as parameters when needed (this is currently used to inject `ScrapingContext`)
  + Validation errors are logged using the `logging` package
- Wrote scraping actions to mirror the capabilities of the previous system
- See `scraping.actions.scrape_actions` for all of the current actions

### CODE REVIEW

**Comment by JeremyEastham** [comment]
### To-Do After Merge:
- Create more scrape actions (scroll, click, follow link, etc)
- Implement regex filtering
- Finish scraping Tesla and scrape more websites

**Comment by JeremyEastham** [comment]
This PR is ready to review but **NOT** ready to merge yet.

File: agtern/common/models.py [hunk 1]
```diff
@@ -1,14 +1,26 @@
...
+class Internship(BaseModel):
     """Models internship details."""
 
     company: str
     title: str
-    year: str
-    period: str
     link: str
+    period: Season = ""
+    year: int = ""
```
Comment (johndpjr) [inline agtern/common/models.py:24]: Setting the default val of year to ""? Shouldn't it be some default int?

Comment (JeremyEastham) [reply agtern/common/models.py]: I changed the default to 0.

**Review by johndpjr — APPROVED** [review]
LGTM! Good job Jeremy this adds lots of value to the project and makes it a lot easier to scrape in the future

**Comment by JeremyEastham** [comment]
### Done:
- Regex support for the `scrape` action
- Nested model parsing for `scrape_action` parameters
- Saving and loading internship data where scrape data does not necessarily match GUI data
- Graceful type coercion with defaults and validation errors (including custom validation)

### To-Do:
- Create more scrape actions (scroll, click, follow link, etc)
- Finish scraping Tesla and scrape more websites
- Display logs from the scraping process when run with the GUI (need inter-process logging queue)
- Since an Internship's year defaults to 0, 0 is being displayed in the GUI
  + I thought about fixing this, but it would probably cause a conflict with the other GUI changes upstream
  + We might consider changing the default back to `""` or `None`
- Export JSON Schema of scrape action parameters for IntelliSense in VS Code/PyCharm?

**Comment by JeremyEastham** [comment]
I fixed the IntelliSense on the ScrapingContext that was not working during the meeting.

Also, I plan to update the docs with a guide on how to write the scraping config and scraping actions.

**Comment by JeremyEastham** [comment]
I've written code that closes #63, but I'll open another PR because we probably need to discuss the implementation regarding the issues outlined in #62.

**Pull request merged**